blockchain/indexers/cfindex.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		return err
+	}
+
+	return dbStoreFilterHeader(dbTx, cfHeaderKeys[wire.GCSFilterExtended],
+		&idx.chainParams.GenesisBlock.Header.PrevBlock, firstHeader)
+}
+
+// storeFilter stores a given filter, and performs the steps needed to
+// generate the filter's header.
+func storeFilter(dbTx database.Tx, block *dcrutil.Block, f *gcs.Filter, filterType wire.FilterType) error {
+	if uint8(filterType) > maxFilterType {
+		return errors.New("unsupported filter type")
+	}
+
+	// Figure out which buckets to use.
+	fkey := cfIndexKeys[filterType]
+	hkey := cfHeaderKeys[filterType]
+
+	// Start by storing the filter.
+	h := block.Hash()
+	var basicFilterBytes []byte
+	if f != nil {
+		basicFilterBytes = f.NBytes()
+	}
+	err := dbStoreFilter(dbTx, fkey, h, basicFilterBytes)
+	if err != nil {
+		return err
+	}
+
+	// Then fetch the previous block's filter header.
+	ph := &block.MsgBlock().Header.PrevBlock
+	pfh, err := dbFetchFilterHeader(dbTx, hkey, ph)
+	if err != nil {
+		return err
+	}
+
+	// Construct the new block's filter header, and store it.
+	prevHeader, err := chainhash.NewHash(pfh)
+	if err != nil {
+		return err
+	}
+	fh := gcs.MakeHeaderForFilter(f, prevHeader)
+	return dbStoreFilterHeader(dbTx, hkey, h, fh[:])
+}
+
+// ConnectBlock is invoked by the index manager when a new block has been
+// connected to the main chain. This indexer adds a hash-to-cf mapping for
+// every passed block. This is part of the Indexer interface.
+func (idx *CFIndex) ConnectBlock(dbTx database.Tx, block, parent *dcrutil.Block, view *blockchain.UtxoViewpoint) error {
+	f, err := builder.BuildBasicFilter(block.MsgBlock())
+	if err != nil && err != gcs.ErrNoData {
+		return err
+	}
+
+	err = storeFilter(dbTx, block, f, wire.GCSFilterRegular)
+	if err != nil {
+		return err
+	}
+
+	f, err = builder.BuildExtFilter(block.MsgBlock())
+	if err != nil && err != gcs.ErrNoData {
+		return err
+	}
+
+	return storeFilter(dbTx, block, f, wire.GCSFilterExtended)
+}
+
+// DisconnectBlock is invoked by the index manager when a block has been
+// disconnected from the main chain.  This indexer removes the hash-to-cf
+// mapping for every passed block. This is part of the Indexer interface.
+func (idx *CFIndex) DisconnectBlock(dbTx database.Tx, block, parent *dcrutil.Block, view *blockchain.UtxoViewpoint) error {
+	for _, key := range cfIndexKeys {
+		err := dbDeleteFilter(dbTx, key, block.Hash())
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, key := range cfHeaderKeys {
+		err := dbDeleteFilterHeader(dbTx, key, block.Hash())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// FilterByBlockHash returns the serialized contents of a block's basic or
+// extended committed filter.
+func (idx *CFIndex) FilterByBlockHash(h *chainhash.Hash, filterType wire.FilterType) ([]byte, error) {
+	var f []byte
+	err := idx.db.View(func(dbTx database.Tx) error {
+		if uint8(filterType) > maxFilterType {
```

Comment:
I'd suggest doing this at the top of the function before opening a database transaction.